### ISSUE TITLE
Allow query tags to be forced inside generated SQL to prevent SQL Server from stripping them from "query text entries"

### ISSUE THREAD

**Issue by jzabroski** [dotnet/efcore#20105]
https://github.com/aspnet/EntityFramework.Docs/blob/master/entity-framework/core/querying/tags.md

Based on helpful discussion with @ajcvickers 

This might be somewhat controversial, but I think many people including @m60freeman and @nickcraver want very similar features: Easy way to scrap boilerplate, reduce dependency on copy-paste errors in log messages, etc.

I think the Query Tags documentation would benefit from a broader example that demonstrates how to use a self-contained data access layer (through a Repository pattern).  The exact size of the sample code could start off small as people give their feedback, but the basic goal should be to give DBAs and engineers a language to discuss how to tag queries with pseudo-exact code locations.

My code sample would look something like that following:

```csharp
class PersonRepository : IRepository<Person> {
  DbContext _dbContext;
  PersonRepository(DbContext dbContext) {
    _dbContext = dbContext;
  }
  private static string GetTypeAndMethodName([CallerMemberName] string callerName = null)
  {
    return $"{typeof(PersonRepository).Name}.{callerName}";
  }
  IQueryableResult<Person> All() { // assume IQueryableResult is an abstraction which blocks adding method chains to an IQueryable<T>
    return DbSet<Person>().TagWith($"{GetTypeAndMethodName()}"); // this will tag the query with "PersonRepository.All"
  }
}
```

**Comment by ajcvickers** [comment]
We discussed this again. For most uses of query tags and for most providers the current behavior is fine. For example, the tags are in logs (which was the most asked for reason to have tags) and we verified that they still show up in the SQL Server profiler. We're also conscious that generating readable queries ("like you would write them") is a goal for EF Core, although it is less important generally than doing something that works.

We think the balance here is to leave the current SQL formatting the same, but to add an opt-in that moves the tags after the first SELECT or at the end of the first line, etc. as discussed above and shown in the gist.

**Comment by robfe** [comment]
I'm very keen to get tags (I'm using caller member name and caller file name) showing up in the query store (and Azure's SQL performance recommendations), really glad i found this thread...

I've found a pretty good way to get leading / trailing comments into the query store: Wrap the whole statement in parens:

```
(
	/*Yes!*/
	SELECT TOP 1 'Does it work with starslash in parens' from sys.tables
)

(
	--Yep!
	SELECT TOP 1'Does it work with starslash in parens' from sys.tables
)


SELECT * FROM sys.query_store_query_text where query_sql_text like '%Does it%' AND query_sql_text not like '%query_store_query_text%'


```

Yields
```
query_text_id	query_sql_text	statement_sql_handle	is_part_of_encrypted_module	has_restricted_text
29	(   /*Yes!*/   SELECT TOP 1 'Does it work with starslash in parens' from sys.tables  )	0x09006C16510497B0E0663D7A79B9EFA7A7120000000000000000000000000000000000000000000000000000	0	0
30	(   --Yep!   SELECT TOP 1'Does it work with starslash in parens' from sys.tables  )	0x0900A4F4734F6B052044F300946D4217C06F0000000000000000000000000000000000000000000000000000	0	0
```

Can anyone else think of a way this will negatively impact query parsing / performance / etc? Aside from the potential that changing the comment will probably break any forced plans

**Comment by ajcvickers** [comment]
@robfe This should be possible now using an `IDbCommandInterceptor` to wrap the generated SQL.

**Comment by jzabroski** [comment]
@ajcvickers @robfe Great teamwork!

**Comment by jzabroski** [comment]
> Can anyone else think of a way this will negatively impact query parsing / performance / etc? Aside from the potential that changing the comment will probably break any forced plans

Some people use the nuget package from Microsoft to parse the abstract syntax tree for t-sql.  In the Microsoft world, that is:

* Microsoft.SqlServer.TransactSql.ScriptDom nuget package - optimized for syntax tree analysis for third party tools like Microsoft DACPAC - only supports .NET Core and may require DacFx as a hidden dependency
* Microsoft.SqlServer.Management.SqlParser.dll - used by the  language service, so it's optimized around those types of tasks (e.g., IntelliSense suggestions, QuickInfo tooltips, diagnostic errors).

In other words, some people try mucking with the SQL syntax tree using basic string manipulation.  It would probably be better to manipulate the full AST and motion the AST back to a plaintext string when done.

This is the only scenario I can think of that could break - that your underlying parser wouldn't expect a top-level enclosing parentheses.  Other possible breaks might be somehow this disabling various peephole optimizations, but I doubt it.  You can use `SET SHOWPLAN XML` to check for any differences in the underlying plan.

**Comment by RobHudson72** [comment]
I am keenly interested in this issue... Has there been any update?  

If I get the gist of the conversation, this requires a change by the SQL Server team to support tags in QueryStore.  Does anyone know if this is on the roadmap?

**Comment by ajcvickers** [comment]
@RobHudson72 This issue is tracking:

> We think the balance here is to leave the current SQL formatting the same, but to add an opt-in that moves the tags after the first SELECT or at the end of the first line, etc. as discussed above and shown in the gist.

**Comment by jzabroski** [comment]
> I am keenly interested in this issue... Has there been any update?
> 
> If I get the gist of the conversation, this requires a change by the SQL Server team to support tags in QueryStore. Does anyone know if this is on the roadmap?

There is a workaround described, where you can implement an IDbCommandInterceptor and wrap your whole SQL Statement in parentheses.  Then it should work, and the expense of possibly affecting other things (e.g., if you use an external tool like SolarWinds DPA, all your query hashes it stores will change, and any work your DBA has put in to give friendly names to those hashes will be lost).

**Comment by RobHudson72** [comment]
@jzabroski , thanks!  Noted, and working on an internal POC... quite excited about how this might work!

**Comment by forik** [comment]
Wrapping an sql statement containing e.g. ORDER clause in the end of a query won't work though.
![image](https://user-images.githubusercontent.com/886162/130125905-3d5f503d-4e1d-43fd-b86b-bbd8a24df238.png)

**Comment by jzabroski** [comment]
@forik Why can't we have nice things?  I confirmed you're right.  The following does work, though (verified on SQL Server v14.0.3048.4):

```
(
	SELECT * FROM
	(
	/*Yes!*/
	SELECT TOP 1 'Does it work with starslash in parens' AS col1 from sys.tables t ORDER BY t.[type]) AS wrapper
)
;
(
	SELECT * FROM
	(
	--Yep!
	SELECT TOP 1'Does it work with starslash in parens' AS col1 from sys.tables t ORDER BY t.[type]) AS wrapper
)

SELECT * FROM sys.query_store_query_text where query_sql_text like '%wrapper%' AND query_sql_text not like '%query_store_query_text%'
```
yields:
```
query_text_id   query_sql_text                                                                                                                                          statement_sql_handle                                                                            is_part_of_encrypted_module  has_restricted_text
251122425	(   SELECT * FROM   (   /*Yes!*/   SELECT TOP 1 'Does it work with starslash in parens' AS col1 from sys.tables t ORDER BY t.[type]) AS wrapper  )	0x0900A08003BAD6C36DDBC5D41E9650CD0D8C0000000000000000000000000000000000000000000000000000	0                            0
251122426	(   SELECT * FROM   (   --Yep!   SELECT TOP 1'Does it work with starslash in parens' AS col1 from sys.tables t ORDER BY t.[type]) AS wrapper  )   	0x0900F151299CBBC41F88B218A669758437C20000000000000000000000000000000000000000000000000000	0               	     0
```

**Comment by forik** [comment]
@jzabroski yeah I also tried that one with wrapping with SELECT * FROM (...) but it feels a little bit too much, doesn't it?

**Comment by jzabroski** [comment]
When you ask a computer scientist if an indirect way of solving a problem is a little bit too much, just know that our benchmark is the 7 layer OSI model and the POSIX file system interface :D  We don't really care much about indirection as long as we can strongly reason about the results of our designs.

But yeah, it would be nice if the SQL Server team just fixed this nagging issue.

**Comment by m60freeman** [comment]
@jzabroski, @ajcvickers @robfe Can anyone provide the code needed to implement the suggested workaround involving IDbCommandInterceptor that puts a comment  in /* class.method */ after the first space or wraps the statement in parenthesis with such a comment? 

Sorry, I'm a SQL DBA, not a C# developer. :-)  I need something I can hand off to an application development team and say "Please include this in our project".

@ajcvickers Is a true built-in feature still being considered (just not for the current release) or will the feature not be implemented at all because there is a workaround?

**Comment by ajcvickers** [comment]
@m60freeman There's still a possibility, but it's low on the priority list. Make sure to vote (👍) for this issue if it's important to you.

**Comment by m60freeman** [comment]
@ajcvickers Can you please provide a link to the specific issue? 

Keeping track of where Microsoft is storing such things over time is something I gave up on a while back. :-)

**Comment by ajcvickers** [comment]
@m60freeman Sorry, I though you were asking whether we would ever implement putting the tags inside SQL in EF Core. That is tracked by this issue. If you're asking about where SQL Server tracks things they might do, then that's as opaque to me as it is to you.

**Comment by m60freeman** [comment]
@ajcvickers Sorry to be a bit slow, but I don't see where I can vote for this Issue either on this page or in the list of Issues. 

My question about whether this was still being considered was because you removed the "consider-for-current-release" label in October.

**Comment by ajcvickers** [comment]
@m60freeman You can vote for this issue by adding a 👍 on the issue description. We can then [sort by number of votes](https://github.com/dotnet/efcore/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). This issue currently has only one vote, while the [top voted issue](https://github.com/dotnet/efcore/issues/4021) has 402 votes, so this issue is pretty far back in the list.

**Comment by m60freeman** [comment]
@ajcvickers I agree this would be a hack, but it is the only currently possible solution. Are you going to communicate with the SQL Server team about this in any way or is there a Microsoft policy that prohibits coordination between the product groups? Because they can't produce a feature that your group will use to resolve this serious disconnect without this being a joint effort.

**Comment by stap123** [comment]
@ajcvickers Is there an open feature request with the SQL team somewhere we can upvote?

**Comment by ajcvickers** [comment]
@scoriani Any thoughts?

**Comment by scoriani** [comment]
What’s exactly the ask here?
Is it to create an end-to-end mechanism to track and troubleshoot performance of individual EF queries from code to query store?
Or is it to just keep comments and spaces before and after a query for sql text in query store?

From: Arthur Vickers ***@***.***>
Date: Thursday, 27 October 2022 at 11:51
To: dotnet/efcore ***@***.***>
Cc: Silvano Coriani ***@***.***>, Mention ***@***.***>
Subject: Re: [dotnet/efcore] Allow query tags to be forced inside generated SQL to prevent SQL Server from stripping them from "query text entries" (#20105)

@scoriani<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fscoriani&data=05%7C01%7CSilvano.Coriani%40microsoft.com%7Cf14c44c956914e022fb108dab800d61a%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638024611000691039%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=qrbwNyQfK%2BYu1xKPwxGRxFxM4oyglpLGn5N3HQz4eHI%3D&reserved=0> Any thoughts?

**Comment by NickCraver** [comment]
@scoriani I think the base level ask would be "a string of text to go with a query", e.g. think of a comment which is where the query was issued from. For us, it's "Path\Files.cs@129" (path, file, line), but it could be any identifying text. The main purpose is to back track a query in SQL (e.g. something running hot) to where it came from in an application.

**Comment by ajcvickers** [comment]
@scoriani The requirement here is a way to track a query all the way from the client to places where that query shows up in SQL Server diagnostics. EF would then use this mechanism. Right now, the only EF can do is add a comment to the query, but then SQL Server strips those comments out unless they are unnaturally embedded within the SQL somewhere.

**Comment by scoriani** [comment]
Got it. Haven't investigated on this for query store yet, but i can imagine we're stripping down comments to avoid storing multiple entries for the same sql text that in a busy system can bloat query store tables but, in this specific use case, it is very likely that a given query will be matched 1:1 with a given comment. I'll reach out to some query store folks and ask for more details.

**Comment by jzabroski** [comment]
@scoriani We're hoping for a corresponding GitHub issue as well, so the trail doesn't run cold.

**Comment by jzabroski** [comment]
:( I personally wish the heads at Microsoft would just shut down SQL Feedback and Visual Studio feedback and just use codeless GitHub repos for issues. Lots of companies do this already, very successfully. Maybe you are friends with some VIPs at Microsoft and can hint at this. It's shocking how bad the search engines are for VS Issues and SQL Feedback. - I don't even want to use these sites, so I barely bother reporting issues unless someone at Microsoft asks.

**Comment by scoriani** [comment]
Ahh... that's way above my pay grade :) but a group of us repeatedly proposed that already over time. Things are slow in large orgs, and VIPs are not necessarily the issue, but we do not give up ;-)

**Comment by jzabroski** [comment]
@scoriani I think my simplest feature request is:

As a database administrator/devops engineer, I want to disable comment stripping from the Query Store, so that I can embed a comment at the beginning of a SQL statement to track queries back to the application they originate from.

@NickCraver I suppose one alternative, that I don't particularly care for, is to use OpenTelemetry .NET's [Enrich](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.SqlClient/README.md#enrich) operation. The main reason I don't like it is more that it looks unnecessarily complex and I have not really played around with it. But there is an EFCore OpenTelemetry plug-in that maybe would be useful to us. https://github.com/open-telemetry/opentelemetry-dotnet/pull/929

**Comment by m60freeman** [comment]
@scoriani We need a way for EF's .TagWith() to be able to "decorate" generated SQL in a way that will remain intact in Query Store (and hopefully the Plan Cache as well). We have proposed something like "SELECT ColumnName /* _CommentFromTagWith_ */ FROM TableName" to work around comment stripping (which ignores this comment placement), but the EF Team feels this is too much of a hack and they are looking for some more elegant/official way to accomplish the goal. 

The goal is to be able to determine from the SQL text in Query Store where the SQL was generated, whether that be Class.Method or Path\File@LineNumber. A DBA sees the SQL text for something that is performing poorly or producing an incorrect result and wants a developer to change it, but currently the developer has no way to find the code that needs to be changed other than by doing some kind of RegEx search through the entire code base, which is inefficient and error prone.

I am guessing that the EF Team would prefer being able to send something like "/* _CommentFromTagWith_ */ SELECT ColumnName FROM TableName", but @ajcvickers would need to clarify that.

Currently in Query Store, we get a value in sys.query_store_query_text.query_sql_text that looks like: "**(@__ToGuid_0 uniqueidentifier)**SELECT ColumnName FROM TableName WHERE (Id = @__ToGuid_0)". Note that the parameter list is "decorating" the SQL. Maybe there could be _CommentFromTagWith_ between another set of delimiters immediately after the parameter list? That might look like "**(@__ToGuid_0 uniqueidentifier)**(_CommentFromTagWith_)SELECT ColumnName FROM TableName WHERE (Id = @__ToGuid_0)". If there is no such "tag", then omit the empty delimiters. Maybe use [] or {} for this instead of (). I have no preference.

The SQL Team and the EF Team must collaborate to find a solution that will be acceptable to both and will work in combination. Having the EF Team add comments that the SQL Team strips out is what we want to avoid.

**Comment by scoriani** [comment]
SQL takes batches of 1 or more statements.  The batch does not strip anything, and the sql_handle is effectively an md5 hash of all of the characters (comments included). That's why in plan cache comments are already there and collected by external tools like SolarWind and such.
 
Query store has a statement sql handle.  It is defined based on the first character of the first token of the statement and the last character of the last token in that statement.  So, comments between statements are ignored for how the query store defines what is a statement.  The statement sql handle is also an md5 hash over the characters.  Adding a space or additional characters like comments makes it look completely different to the system from a caching perspective (batch) or plan forcing perspective (statement or batch).  So, you don’t want adding text to a comment block to undo all of the history and forcing in the query store, so it isn’t counted.  This would have way to many implications and impact on a number of features.
 
Also, adding multiple entries in sys.query_store_query_text for the same query is not viable as memory management with query store is a major concern for busy systems or in Azure SQL where resources are generally more constrained.

Based on the conversations with some team members, it's very unlikely that this behaviour is going to change. General recommendation would be to inject tags/comments within the statement text instead.

**Comment by jzabroski** [comment]
@ajcvickers Is this a stalemate?

**Comment by ajcvickers** [comment]
I think this discussion has become fixated on comments. The requirement as I see it and stated above is:

> The requirement here is a way to track a query all the way from the client to places where that query shows up in SQL Server diagnostics. EF would then use this mechanism.

I went on to say:

> Right now, the only thing EF can do is add a comment to the query, but then SQL Server strips those comments out unless they are unnaturally embedded within the SQL somewhere.

The fact that SQL Server sees comments as something that its reasonable, and sometimes even necessary, to strip out indicates that comments are not the correct way to do this. There should instead be a different, end-to-end mechanism that allows a query to be traced from source code to diagnostics.

**Comment by jzabroski** [comment]
@scoriani Is it possible to go back to the team once more and request some end-to-end mechanism that allows a query to be traced from source code to diagnostics?  I now see why @NickCraver suggested "a string of text to go with a query" after reading @ajcvickers 's reply directly above.

Also, based on your explanation of the internals, it sounds like it is not even feasible to have a Query Store option to not strip out leading comments, since Query Store only knows sql statement handles. e.g., it sounds like the following would not be possible:

```sql
ALTER DATABASE AdventureWorks2014
SET QUERY_STORE = ON   
    (  
    STRIP_COMMENTS = OFF
    );
```

The reason I was asking is cost to implement a feature. But I agree that "a string of text to go with a query" is probably the best way to provide an "end-to-end mechanism that allows a query to be traced from source code to diagnostics."

**Comment by m60freeman** [comment]
@scoriani Adding an additional NVARCHAR column to sys.query_store_query_text to hold the TagWith data from EF would be fine with me, but there would have to be a way to EF to send that information in such a way that Query Store would pick it up. Ideally, it would be something that would also get picked up by the plan cache as part of the batch, but if not it would give people another reason to enable Query Store. 

Otherwise it looks like the "hack" of having EF inject a comment just after the first token of the statement is the only thing that will work. This would change the md5 hash (once), but I really don't have a problem with that. It would just have to be documented in the new EF feature that enabling it will cause a new query_hash to be generated and there would be no way to link the old and new hashes.

**Comment by scoriani** [comment]
Sounds to me that there's the need for much more detailed framing/scoping for such a feature, including key use cases and scenarios.
Is it a single query that needs to be traced end to end? Or is it a client command/unit of work that may generate multiple queries in a single or multiple batches? 
Should we trace a single query execution or aggregated metrics? Remember Query Store only aggregated query execution stats over a time period, not individual query executions.
Why EF and not underlying client drivers and other data access frameworks? What about integrations with other observability tools/frameworks/solutions like OpenTelemetry, Dynatrace, Datadog, Redgate, etc.? 
While i can see the value of such a feature, it will impact many areas inside and outside the engine and will need to be properly evaluated and prioritised together with other product improvements.

**Comment by m60freeman** [comment]
@scoriani The main issue is the EF/LINQ generates queries that are often so long and deeply nested that their text is truncated before even getting to a WHERE clause. Given a FROM and a WHERE, application developers have much less work to do to find where a query is generated. I am not aware of other frameworks that produce code so convoluted and obscured by unhelpful aliases. No one in this thread has asked for anything in the way of metrics other than what is already provided. We just need to know where in the code a query originated. 

Having this information available in the plan cache (a new tag in the plan XML and a new column in sys.dm_exec_sql_text?) would make it available to monitoring tools that capture from there and not Query Store. 

Speaking for myself, I would much rather see a narrowly-defined feature that can be implemented relatively simply and quickly than something that will handle every possible use case. The information we need in SQL Server is clear from reading this thread. The tricky part is defining how to get it there in a way that the EF team is willing to provide.

**Comment by ajcvickers** [comment]
@scoriani 

> Is it a single query that needs to be traced end to end? Or is it a client command/unit of work that may generate multiple queries in a single or multiple batches?

Single query.

> Should we trace a single query execution or aggregated metrics? Remember Query Store only aggregated query execution stats over a time period, not individual query executions.

That's a good question--maybe @NickCraver has an idea of what would be best here?

> Why EF and not underlying client drivers and other data access frameworks? What about integrations with other observability tools/frameworks/solutions like OpenTelemetry, Dynatrace, Datadog, Redgate, etc.?

This should definitely _not_ be an EF-specific thing. It should be something usable by any client. EF would make use the of the underlying mechanism just like any other client.

> While i can see the value of such a feature, it will impact many areas inside and outside the engine and will need to be properly evaluated and prioritised together with other product improvements.

I totally understand that the value of implementing this might not be worth the cost it takes to implement, both in terms of resources, and in terms of impact on the code or performance. That's fine, and a statement as such from the SQL team would put this issue to bed as not something that SQL Server supports.

**Comment by jzabroski** [comment]
I'd just chime in to say that I think @m60freeman 's point that a lot of the tracing value here relates to impossibly long SQL queries generated by ORMs to be fairly accurate.

Thus, I don't think aggregated metrics matter that much. The only use case for aggregated metrics is when you have nested transactions, because the inner transaction does not record wait types and yet the outer transaction can cause blocking due to session-level resource tracking. Thus, tools like SolarWinds DPA show a 0 cost on things that actually have a high aggregate cost. But, that's what session-based diagnostic tools like SQL Sentry Profiler help with already. Aggregate metrics is more or less a solved problem in my mind. Could it be smoother? Yes. Do I need it to be smoother? No.

User stories I've cared about in the past:

1. SELECT FROM statements without a where clause; root cause is C# developers don't understand LINQ query execution semantics and typically arises when they try joining a SQL table to an in-memory collection. _Finding_ the bad query is actually fairly hard to do.
2. Somebody writes an overly complicated `GROUP BY` clause that uses object equality, rather than simpler field equality (perhaps taking advantage of a unique constraint) and possibly using a `cross apply` instead of a `group by`. I need to find where that query is.
3. SQL Server suddenly chooses a bad query plan for an EF query. I want to be able to peg the old plan ASAP. (This is a really high value item for me. If Postgres had such a feature, I would never go back to SQL Server... especially now that Amazon MariaDB has a T-SQL engine)

For 1 and 2, one way EF Core team could be helpful is in strictly promoting the best practice of repositories that don't leak `IQueryable`. Code bases have an insane amount of technical debt when you have bad code + layering violations, because it makes it so time consuming to track this non-sense down. As I stated [in 2020](https://github.com/dotnet/efcore/issues/20105#issuecomment-587173741), I don't leak `IQueryable` any more because of the maintenance nightmare it becomes tracking down poor performing SQL. To draw an analogy, the ASP.NET Core team outlawed certain types of Dependency Injection patterns, like SimpleInjector/NInject allowing use of `Func<T>` to create ambient object instances from an implicit "current scope". It would be a net benefit if EF Core query examples didn't just use a `DbContext` but used a repository abstraction that didn't leak IQueryable, but transformed that into something representing a "promise to return a collection" (VectorResultSet, whatever you want to call it).
